bvm/ballerina-core/src/main/java/org/ballerinalang/model/util/serializer/JsonDeserializer.java

```diff
@@ -0,0 +1,442 @@
+     * @param target   type of which the object to be created.
+     * @return instance of given type.
+     */
+    private Object createInstance(BMap<String, BValue> jsonNode, Class<?> target) {
+        if (Enum.class.isAssignableFrom(target)) {
+            return createEnumInstance(jsonNode);
+        }
+        if (jsonNode.hasKey(JsonSerializerConst.TYPE_TAG)) {
+            BValue typeNode = jsonNode.get(JsonSerializerConst.TYPE_TAG);
+            String type = typeNode.stringValue();
+            if (type.equals(JsonSerializerConst.ARRAY_TAG)) {
+                return createArrayInstance(jsonNode);
+            }
+            return getObjectOf(type);
+        }
+        return getObjectOf(target);
+    }
+
+    private Object createArrayInstance(BMap<String, BValue> jsonNode) {
+        BString ct = (BString) jsonNode.get(JsonSerializerConst.COMPONENT_TYPE);
+        String componentType = ct.stringValue();
+
+        BInteger bSize = (BInteger) jsonNode.get(JsonSerializerConst.LENGTH_TAG);
+        int size = (int) bSize.intValue();
+
+        Class<?> clazz = findClass(componentType);
+        if (clazz != null) {
+            return Array.newInstance(clazz, size);
+        }
+        throw new BallerinaException("Can not create array instance of: " + componentType + "[]");
+    }
+
+    @SuppressWarnings("unchecked")
+    private Object createEnumInstance(BMap jsonNode) {
+        String enumName = jsonNode.get(JsonSerializerConst.PAYLOAD_TAG).stringValue();
+        int separatorPos = enumName.lastIndexOf(JsonSerializerConst.ENUM_SEPERATOR);
+        String type = enumName.substring(0, separatorPos);
+        String enumConst = enumName.substring(separatorPos + 1);
+
+        Class enumClass = instanceProvider.findInstanceProvider(type).getTypeClass();
+        return Enum.valueOf(enumClass, enumConst);
+    }
+
+    private Object getObjectOf(Class<?> clazz) {
+        String className = ObjectHelper.getTrimmedClassName(clazz);
+        return getObjectOf(className);
+    }
+
+    private Object getObjectOf(String type) {
+        TypeInstanceProvider typeProvider = instanceProvider.findInstanceProvider(type);
+        return typeProvider.newInstance();
+    }
+
+    @SuppressWarnings("unchecked")
+    private Object deserializeReflectively(Object instance, BMap<String, BValue> payload) {
+        setFields(instance, payload);
+        return instance;
+    }
+
+    private void setFields(Object target, BMap<String, BValue> fieldMap) {
+        Objects.requireNonNull(target);
+        Class<?> targetClass = target.getClass();
+        HashMap<String, Field> allFields = ObjectHelper.getAllFields(targetClass, 0);
+
+        for (String fieldName : fieldMap.keys()) {
+            if (fieldName.equals(JsonSerializerConst.ID_TAG)) {
+                // it's a metadata entry.
+                continue;
+            }
+            Field field = allFields.get(fieldName);
+            if (field == null) {
```

Comment:
use `allFields.containsKey()`
check in all places..

Reply:
That way we will have to compute the hashMap lookup twice, one for containsKey and then for get.